### ISSUE TITLE
startService/stopService should cancel pending callLater for clarity in the finger tutorial

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3375]
|<img alt="helo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| helo reported|
|-|-|
|Trac ID|trac#3375|
|Type|defect|
|Created|2008-07-30 18:20:14Z|
|Branch|https://github.com/twisted/twisted/tree/callater-finger-3375|

The definitions of `startService` and `stopService` at [doc/core/howto/tutorial/listings/finger/finger14.tac#L28](doc/core/howto/tutorial/listings/finger/finger14.tac#L28) should probably be propagated into the later fingerNN.tac files to avoid confusion.

Attachments:

* [finger-3375.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a18/a18271c052c703b93a7970cb0c1f46134f676e1e/5968c44439c1c671b8cbe23810f0c4af0e0c3fae.patch) (5788 bytes) - added by thijs on 2011-01-16 21:26:56Z - against r30465

<details><summary>Searchable metadata</summary>

```
trac-id__3375 3375
type__defect defect
reporter__helo helo
priority__lowest lowest
milestone__ 
branch__branches_callater_finger_3375 branches/callater-finger-3375
branch_author__thijs thijs
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1217442014000000 1217442014000000
changetime__1295705463000000 1295705463000000
version__None None
owner__ 
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @thijstriemstra|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_5"></a>|@cyli commented|
|-|-|

Reading through the tutorial, the [step at which finger14.tac is linked from](http://twistedmatrix.com/documents/current/core/howto/tutorial/protocol.html#auto3) has this snippet of text:

  Here we override the standard `startService` and `stopService` hooks in the Finger service, which is set up as a child service of the application in the last line of the code. `startService` calls `_read`, the function responsible for reading the data; `reactor.callLater` is then used to schedule it to run again after thirty seconds every time it is called. `reactor.callLater` returns an object that lets us cancel the scheduled run in `stopService` using its `cancel` method.

So it seems like both `startService` and `stopService` were introduced in this step, but neither gets propagated into the later tac files.  This ticket is for propagating the `stopService` calls, but the `startService` calls are also not propagated. `_read` is instead called from `__init__`, despite the fact that the tutorial does not mention any reason for this change in [the next step](http://twistedmatrix.com/documents/current/core/howto/tutorial/protocol.html#auto4) or any subsequent step.

Since this ticket is already propagating `stopService` to the subsequent fingerNN.tac files, shouldn't `startService` also be propagated?

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_6"></a>|@cyli set owner to @thijstriemstra|
|-|-|

As per the discussion in the channel, propagating `startService` should also be part of this ticket.  This would also mean removing the call to `_read` in `__init__`, since it will be called in `startService` instead.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_7"></a>|@thijstriemstra set status to `assigned`|
|-|-|

Thanks for the review, I'll look into it.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_8"></a>|@thijstriemstra commented|
|-|-|

(In [30545]) Branching to 'callater-finger-3375'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_9"></a>|@thijstriemstra commented|
|-|-|

(In [30546]) Apply finger-3375.patch, refs [#3375](https://github.com/twisted/twisted/issues/3375)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_10"></a>|@thijstriemstra commented|
|-|-|

updating ticket title.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra commented|
|-|-|

(In [30547]) Propagate startService and move _read call, apply coding standard. refs [#3375](https://github.com/twisted/twisted/issues/3375)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_12"></a>|@thijstriemstra commented|
|-|-|

(In [30548]) added news file, refs [#3375](https://github.com/twisted/twisted/issues/3375)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_14"></a>|@cyli set owner to @thijstriemstra|
|-|-|

Thanks for working on this thijs!  Looks good, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_15"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [30555]) Merge callater-finger-3375: Added the stopService and startService methods to all finger example files.

Author: thijs
Reviewer: cyli
Fixes: [#3375](https://github.com/twisted/twisted/issues/3375)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
